mlos_core/mlos_core/tests/optimizers/optimizer_test.py

```diff
@@ -298,3 +298,69 @@ def test_optimizer_type_defs(optimizer_class: Type[BaseOptimizer]) -> None:
     """
     optimizer_type_classes = {member.value for member in OptimizerType}
     assert optimizer_class in optimizer_type_classes
+
+
+@pytest.mark.parametrize(('optimizer_type', 'kwargs'), [
+    # Default optimizer
+    (None, {}),
+    # Enumerate all supported Optimizers
+    *[(member, {}) for member in OptimizerType],
+    # Optimizer with non-empty kwargs argument
+])
+def test_mixed_input_space_types(optimizer_type: Optional[OptimizerType], kwargs: Optional[dict]) -> None:
+    """
+    Toy problem to test the optimizers.
+    """
+    max_iterations = 10
+    if kwargs is None:
+        kwargs = {}
+
+    def objective(point: pd.DataFrame) -> pd.Series:
+        # mix of hyperparameters, optimal is to select the highest possible
+        ret: pd.Series = point["x"] + point["y"]
+        return ret
+
+    input_space = CS.ConfigurationSpace(seed=2169)
+    # add a mix of numeric datatypes
+    input_space.add_hyperparameter(CS.UniformIntegerHyperparameter(name='x', lower=0, upper=5))
+    input_space.add_hyperparameter(CS.UniformFloatHyperparameter(name='y', lower=0.0, upper=5.0))
+
+    if optimizer_type is None:
+        optimizer = OptimizerFactory.create(
+            parameter_space=input_space,
+            optimizer_kwargs=kwargs,
+        )
+    else:
+        optimizer = OptimizerFactory.create(
+            parameter_space=input_space,
+            optimizer_type=optimizer_type,
+            optimizer_kwargs=kwargs,
+        )
+
+    with pytest.raises(ValueError, match="No observations"):
+        optimizer.get_best_observation()
+
+    with pytest.raises(ValueError, match="No observations"):
+        optimizer.get_observations()
+
+    for _ in range(max_iterations):
+        suggestion = optimizer.suggest()
+        assert isinstance(suggestion, pd.DataFrame)
+        assert (suggestion.columns == ['x', 'y']).all()
+        # Build suggestion mapping to cooperate with Configuration, note that
+        # doing a .iloc[0].to_dict() will cause pandas convert all numeric types
+        # to float64
+        tmp_suggest = {'x': suggestion['x'].values[0], 'y': suggestion['y'].values[0]}
```

Comment:
A little concerned about this part.

Is the intent behind `tmp_suggest` for the test to avoid particular code paths that might do type conversions we don't want so that we can check that the types were correct when not doing that?  If so, can you amend the comment to document that a bit more?

I guess what I'm worried about is for callers to have to be concerned about these types of things.  We should probably try and handle it internally as much as possible.

In principle, I think it's probably an error if we can't immediately do `register` with the value returned from `suggest`.

Reply:
Yes so if we take a look a previous test at [`optimizer_tests.py:93`](https://github.com/microsoft/MLOS/blob/main/mlos_core/mlos_core/tests/optimizers/optimizer_test.py#L93), this code path will fail if we utilize a mix of input types (namely floats and integers). This is due to the fact that `pd.DataFrame.iloc` will always promote data types. We can fix this by calling `.astype('O')` before any `iloc` call, but again that would mean the caller would have to be worried about this type casting.

For now I can remove `tmp_suggest` and instead replace the input with `suggestion.astype('O').iloc[0].to_dict()`

Reply:
I like this new version a little better.  Consistent with the code used elsewhere as well.  Thanks!

Reply:
As an aside, because of these weird behaviors of pandas, I'm not certain it was necessarily the right data structure to use as our APIs.  The original thought was that it was better from a data science user perspective, but mlos_bench later took a different tact of using dicts more frequently so we end up converting for the user quite a bit anyways.  Automatic type conversions behind the scenes based on your access method may be convenient in some circumstances, but not others.

I think at least in the common case, mlos_bench is the main user of mlos_core, so we can potentially mask some of it.

Anyways, just some backstory.  Very much appreciate you catching this.  Thanks!